lib/vector-buffers/src/topology/channel/limited_queue.rs

```diff
@@ -213,462 +187,441 @@ pub struct LimitedReceiver<T> {
     inner: Inner<T>,
 }
 
-impl<T: Bufferable> LimitedReceiver<T> {
+impl<T: Send + 'static> LimitedReceiver<T> {
     /// Gets the number of items that this channel could accept.
     pub fn available_capacity(&self) -> usize {
         self.inner.limiter.available_permits()
     }
 
-    pub fn poll_next(&mut self, cx: &mut Context<'_>) -> Poll<Option<T>> {
+    pub async fn next(&mut self) -> Option<T> {
         loop {
-            if let Some((permit, item)) = self.inner.data.pop() {
-                // We got an item, woohoo! Now, drop the permit which will properly free up permits
-                // in the semaphore, and then also try to notify a pending writer.
-                drop(permit);
-                self.inner.write_waker.as_ref().notify_one();
-
-                return Poll::Ready(Some(item));
+            if let Some((_permit, item)) = self.inner.data.pop() {
+                return Some(item);
             }
 
             // There wasn't an item for us to pop, so see if the channel is actually closed.  If so,
             // then it's time for us to close up shop as well.
             if self.inner.limiter.is_closed() {
-                return Poll::Ready(None);
+                return None;
             }
 
             // We're not closed, so we need to wait for a writer to tell us they made some
             // progress.  This might end up being a spurious wakeup since `Notify` will
-            // store up to one wakeup that gets consumed by the next call to `poll_notify`,
-            // but alas.
-            ready!(self.inner.read_waker.poll_notify(cx));
+            // store a wake-up if there are no waiters, but oh well.
+            self.inner.read_waker.notified().await;
         }
     }
 
-    pub fn close(&mut self) {
-        self.inner.limiter.close();
-    }
-}
-
-impl<T: Bufferable> Stream for LimitedReceiver<T> {
-    type Item = T;
-
-    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<T>> {
-        Pin::into_inner(self).poll_next(cx)
-    }
-}
-
-impl<T> Clone for LimitedSender<T> {
-    fn clone(&self) -> Self {
-        self.sender_count.fetch_add(1, Ordering::SeqCst);
-
-        Self {
-            inner: self.inner.clone(),
-            sender_count: Arc::clone(&self.sender_count),
-            slot: None,
-        }
+    pub fn into_stream(self) -> Pin<Box<dyn Stream<Item = T> + Send>> {
+        let mut receiver = self;
+        Box::pin(stream! {
+            while let Some(item) = receiver.next().await {
+                yield item;
+            }
+        })
```

Comment:
It probably doesn't matter, but I'm slightly curious if there's more overhead here than a manual `Stream` impl (which doesn't seem like it'd need to be terribly complex).

Reply:
From what I've seen of the generated code, it's all pretty close.

The main difference is that `async-stream` uses some thread-local `Cell` stuff to let the actual `async` guts -- the generator -- push the yielded item somewhere that the `Stream` implementation it generates can consume it, like a global reusable oneshot, almost.

Technically, the overhead is super low because it's a thread-local and it's using `Cell`, not any sort of actual async channel mechanism... but it's potentially _more_ efficient since it only has to create the async generator future once, vs what I ended up doing for `BufferReceiverStream`, where we do have `ReusableBoxFuture` so we don't have to reallocate but we do need to recreate the future each time as we pass the `BufferReceiver` back and forth.

Overall, it strikes me as a wash but in the long-term, it's probably worth doing a manual `Stream` wrapper because this is what things like `tokio-util` end up doing for their own `async`-native channels/primitives... and that likely isn't a coincidence.